### ISSUE TITLE
Date picker announces days of the week as letters (i.e., "Th") JAWS or NVDA as user moves through days

### ISSUE THREAD

**Issue by DavidMacDonald** [w3c/aria-practices#1570]
The abbr element is not well supported in this scenario in a header column of a table. It never has been and I doubt it will be.

Some screen reader users may be fine with hearing these letters instead of the actual day but I am assuming many would prefer to hear the day spoken as they arrow through the day numbers.
WCAG requires solutions to be accessibility supported
I think a more robust solution might be to hide the letters from a screen reader and provide an offscreen full days.
### Current code
```
      <thead>
        <tr>
          <th scope="col" abbr="Sunday">
            Su
          </th>
          <th scope="col" abbr="Monday">
            Mo
          </th>
```
### suggested code
```
<thead>
        <tr>
          <th scope="col">
          <span aria-hidden="true">Su</span><span class="visuallyhidden"> Sunday</span>
          </th>
          <th scope="col" >
           <span aria-hidden="true">Mo</span><span class="visuallyhidden"> Monday</span>
          </th>

.visuallyhidden { height: 1px; width: 1px; position: absolute; overflow: hidden; top: -10px; }
```

**Comment by JAWS-test** [comment]
The abbr attribute works within tables, at least with Firefox and IE 11, but not with Chrome and Edge. It does not work inside grids and I did not expect it to work in grids. By the way, the elegant solution for grids is aria-label (instead of the complicated solution with hidden text):

``` 
<th aria-label=Monday>Mo</th>
```

**Comment by DavidMacDonald** [comment]
arIa-label works well with JAWS and VoicOver on Chrome but not on NVDA.

**Comment by mcking65** [comment]
The APG naming guidance strongly discourages hiding visible text or using aria-label on elements that get their name from content and include visible text content. These are generally risky techniques.

In this narrow case, one could argue there is benefit with only marginal downside to using aria-label. However, in general, overriding the content of a column header or hiding text content in a column header would be very problematic, especially if the header includes interactive content.

We discussed this issue when developing the example. The conclusions were:
1. The abbr technique is well supported by browsers and the HTML spec.
2. When correctly supported, the abbr technique is a superior solution to alternative techniques; it provides significant user benefit not achievable any other way.
3. The APG needs examples that demonstrate this technique in order for the aria-at project to drive improved support.
4. The downsides of lack of support in this example are insignificant.

Net, this is a valuable technique that can solve significant verbosity and presentation problems for screen reader users. The combobox date picker is on the short list of widgets that may have aria-at tests developed before the end of the year. So, so I hope progress is imminent.

Closing this issue as not for practices because we anticipate resolution via aria-at.

**Comment by JAWS-test** [comment]
@mcking65 

> The abbr technique is well supported by browsers and the HTML spec.

abbr is supported within HTML and is specified for HTML, but not automatically for ARIA tables like role=table and role=grid, right?

**Comment by DavidMacDonald** [comment]
It's 2025, six years since the bug was filed. As of today, JAWS does not read the ABBR and announces the dates as mon, tu, we, th etc. 
One of the things we've always tried to do in WCAG is balance reality with aspirations. I would like to recommend we revisit the "out of scope" label this bug received 
https://github.com/orgs/w3c/projects/140/views/1?filterQuery=label%3A%22Out+of+APG+Scope%22
WCAG requires only accessibility supported means of meeting the standard.
https://www.w3.org/TR/WCAG22/#dfn-accessibility-supported

Let's provide something that works using offscreen text as suggested and place a note that it can be changed to ABBR when that is supported by

**Comment by jugglinmike** [comment]
The ARIA Authoring Practices (APG) Task Force just discussed `Issue 3319: Use of ABBR in date picker`. (I'm copying it here because the substance of the discussion is more relevant to this discussion thread)

<details><summary>The full IRC log of that discussion</summary>
&lt;jugglinmike> Topic: Issue 3319: Use of ABBR in date picker<br>
&lt;jugglinmike> github: https://github.com/w3c/aria-practices/issues/3319<br>
&lt;jugglinmike> Matt_King: We started talking about this last week<br>
&lt;jugglinmike> Matt_King: We had a pretty small crew, though, and we didn't come up with a clear next step<br>
&lt;jugglinmike> Matt_King: Really, in issue 3319, David is just asking us to re-open a prior issue that's related to ABBR<br>
&lt;jugglinmike> Matt_King: Five years ago, I boldly said that it's imminent that we're going to test the date picker in 2020<br>
&lt;jugglinmike> Matt_King: David is probably right that it is not imminent<br>
&lt;jugglinmike> Matt_King: What we use ABBR for in the date picker combo box is that the grid has column headings for the days of the week--the content in the column headings is just two character (e.g. "Mo" for "Monday")<br>
&lt;jugglinmike> Matt_King: The column headers have ABBR, which specify the full form of the day of the week<br>
&lt;jugglinmike> Matt_King: When you do that in regular tables, some screen readers (I don't know how many) will read the ABBR across the row (eg "Monday 1, Tuesday 2, Wednesday 3")<br>
&lt;jugglinmike> Matt_King: However, in this implementation in grid, none of the screen readers are doing that<br>
&lt;jugglinmike> Matt_King: Instead, you hear something like "Mo 1 Tue 2 We 2"<br>
&lt;jugglinmike> Matt_King: We could use aria-label. That's one solution<br>
&lt;jugglinmike> Matt_King: Another solution is not to use two-character abbreviations<br>
&lt;jugglinmike> Matt_King: What should the ARIA best-practice be for a date-picker grid like this?<br>
&lt;jugglinmike> Matt_King: First, let's answer the basic question: can we re-open 1570?<br>
&lt;jugglinmike> CurtBellew: I think so. This hits us everyone once in a while. There's not a good way to announce the un-abbreviated version of the day of the week<br>
&lt;jugglinmike> Matt_King: Okay, I've re-opened the issue<br>
&lt;Daniel> q+<br>
&lt;jugglinmike> ack Daniel<br>
&lt;jugglinmike> Daniel: Last week in ARIA we had a discussion about ABBR. It was on the other side--saying that ABBR should not be part of the accessible name (which I would agree with)<br>
&lt;jugglinmike> Daniel: In this case, as long as it's on the description, that's fine with me. But we should be careful with the message that we're communicating. That ABBR is a secondary piece of information, and in this case, the number of the day should be the accessible name<br>
&lt;jugglinmike> Matt_King: Column headers and row headers do not name cells. They are not labels; they're an association<br>
&lt;jugglinmike> Matt_King: This is a really complex issue!<br>
&lt;jugglinmike> Matt_King: The content of the content-header for Tuesday of this month is "2 Tu". If you were to put a label on the column header, I thin that would hide the "t u"<br>
&lt;jugglinmike> Matt_King: If that's case, it would be a fix, but it would have the side-effect of screen readers not being able to know that it's actually just a "T u" there and not the full word "Tuesday"<br>
&lt;jugglinmike> Matt_King: Hiding text that's on the screen has always been something that I think we should approach pretty gingerly<br>
&lt;jugglinmike> Daniel: From a screen-reader user's perspective, is that something we really want?<br>
&lt;jugglinmike> Matt_King: My bigger concern here isn't this specific use-case, but it is the more general use-case of: what do you do when you have a very verbose column header (with a lot of content), but then you're reading across in the table below, and you have to hear the whole header every time. The ABBR element was a good way to abbreviate that<br>
&lt;jugglinmike> Daniel: That's a very good point. I wonder if this is more screen-reader configuration (on whether to read the header after reading the cell or before reading it)<br>
&lt;jugglinmike> Matt_King: I switch it around fairly often. It's a lot of work to switch it, but sometimes you really want it before, and other times you really want it after<br>
&lt;jugglinmike> Matt_King: But it still seems that there is a problem that needs a solution, and it seems like ABBR was doing exactly what we want in that case. It reads the content of the header cell, but it doesn't force you to hear the entire content when you are reading cells that are not the header cell<br>
&lt;jugglinmike> Matt_King: That's why I suggested using ABBR here, to kind of push this as the solution<br>
&lt;jugglinmike> Daniel: Yeah, I don't mean to push hard, but I did want to share another perspective with the group<br>
&lt;jugglinmike> Matt_King: The title of issue 1570 is "date picker announces days of the week as letters". It sounds like Daniel is saying that it is a problem but it isn't a big deal<br>
&lt;jugglinmike> Daniel: Right. We're also using a shorthand, but this isn't technically what ABBR was meant to do<br>
&lt;jugglinmike> Daniel: In this particular case, you observe it correctly, but that may not be so in other cases<br>
&lt;jugglinmike> Matt_King: The actual purpose of ABBR is to provide the expansion<br>
&lt;jugglinmike> Matt_King: so when we have ARIA, the ABBR tag would include the content "Accessible Rich Internet Application"<br>
&lt;jugglinmike> Daniel: Right, and for "Tu", you would have "Tuesday" in the abbreviation<br>
&lt;jugglinmike> Matt_King: The code uses an "abbr" attribute, not an "abbr" element<br>
&lt;Jem> https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/abbr<br>
&lt;jugglinmike> Matt_King: In light of that, I don't know whether what we're doing here is to-spec or not to-spec<br>
&lt;jugglinmike> Matt_King: If anyone wants to do some investigation about whether what we're doing here is in-line with the specification, then that would be helpful<br>
&lt;jugglinmike> Adam_Page: It is a legitimate use<br>
&lt;jugglinmike> Adam_Page: I just checked the specification<br>
&lt;Adam_Page> https://html.spec.whatwg.org/multipage/tables.html#attr-th-abbr<br>
&lt;jugglinmike> Zakim, end the meeting<br>
</details>
